src/sema/function_annotation.rs

```diff
@@ -1,18 +1,131 @@
 // SPDX-License-Identifier: Apache-2.0
 
 use super::{
-    ast::{ConstructorAnnotation, Diagnostic, Namespace, Type},
+    ast::{ConstructorAnnotation, Diagnostic, Expression, Function, Namespace, Type},
     diagnostics::Diagnostics,
-    expression::{expression, ExprContext, ResolveTo},
+    expression::{expression, hex_number_literal, number_literal, ExprContext, ResolveTo},
     unused_variable::used_variable,
     Symtable,
 };
 use crate::Target;
+use num_bigint::BigInt;
+use num_traits::{One, ToPrimitive};
 use solang_parser::pt::{self, CodeLocation};
 
+/// Resolve the prototype annotation for functions (just the selector)
+pub fn function_prototype_annotations(
+    func: &mut Function,
+    annotations: &[&pt::Annotation],
+    ns: &mut Namespace,
+) {
+    let mut diagnostics = Diagnostics::default();
+
+    for annotation in annotations {
+        match annotation.id.name.as_str() {
+            "selector" => function_selector(func, annotation, &mut diagnostics, ns),
+            _ if !func.has_body => {
+                diagnostics.push(Diagnostic::error(
+                    annotation.loc,
+                    format!(
+                        "annotation {} not allowed on {} with no body",
```

Comment:
Why don't you check this in `fn function_selector` by doing `func.has_body()`?

Reply:
`function_selector` handles `@selector(..)` annotations. This check ensures there are no other annotations (for functions without a body). I don't think that belongs in `function_selector()`

Reply:
Ah, got it. Why can't this go to `function_body_annotations` then?

Reply:
It looks like `_ if !func.has_body => ` can go there.

Reply:
I've added a comment explaining this